### ISSUE TITLE
Cannot make TFLMS work with a model/data using up slightly more than my current GPU memory

### ISSUE THREAD

**Issue by fcunilim** [IBM/tensorflow-large-model-support#30]
I compiled your code successfully on Ubuntu 18.04 LTS (real system, not a VM) against the v2.1.0 tag, as stated in your instructions. I very, very carefully followed your guidelines and the ones for building TensorFlow in general. I compiled with XLA on, arch=native, compute capability 6.1, matching my graphics card.

My system is the following:

- core i5 6400 Skylake CPU with 4 threads
- 16GB of system memory
- nVidia GTX 1080 with 8GB

I used CUDA 10.2 and installed the driver provided with it (440.30).
Note my Ubuntu system is freshly installed. CUDA works fine.

Now my program.

The UNet I am using is pretty much mapped on this one : https://github.com/napolar/simple_unet, except I use a 3D version, that's all.

I use the network to segment medical data. With 8GB of memory, my GPU can accept crops of about 96x96x96 in size, without having a memory overflow.
When raising the crop size to 128x128x128, the data can no longer reside in GPU memory.

I decided to try my program with my freshly generated TFLMS build (wheel file), in a clean virtual environment.

Before launching my dev environment, I started with a:

`export TF_GPU_HOST_MEM_LIMIT_IN_MB=16384`

as advised in your documentation. Again, note that 96x96x96 crops do fit in memory in my case, while 128 crops don't.

First, I left TFLMS disabled, to test my TF build. Memory overflowed as expected, a detailed memory output was generated, and I got a TFLMS message in the console output saying enabling TFLMS could make TF load the model just fine.

So I enabled it right after the 'import tensorflow' line, as stated in the doc.

When running the program again with TFLMS enabled, I get the following error:

```
E tensorflow/stream_executor/cuda/cuda_driver.cc:863] failed to alloc 8589934592 bytes on host: CUDA_ERROR_OUT_OF_MEMORY: out of memory
W ./tensorflow/core/common_runtime/gpu/gpu_host_allocator.h:44] could not allocate pinned host memory of size: 8589934592
E tensorflow/stream_executor/cuda/cuda_driver.cc:863] failed to alloc 7730940928 bytes on host: CUDA_ERROR_OUT_OF_MEMORY: out of memory
W ./tensorflow/core/common_runtime/gpu/gpu_host_allocator.h:44] could not allocate pinned host memory of size: 7730940928
```

What have I done wrong?

**Comment by smatzek** [comment]
Does your full error message contain a message like this?
`Allocator (gpu_host_bfc) ran out of memory trying to allocate` ... Note the `gpu_host_bfc` allocator. I believe your message will have this in it. The `gpu_host_bfc` allocator is the allocator of system/CPU memory, not GPU memory.

LMS swaps inactive tensors from GPU memory into system memory. Given the way TensorFlow's memory allocators work, this is a "CUDA pinned host memory" area which also means the operating system cannot page those system memory pages out to its HDD / SDD swap space.

What I believe is happening is that your model is large enough that more than ~16GB of data is being swapped out of the GPU to system memory and you are running out of the system memory space. I say "about 16GB" because your entire system has 16GB of CPU memory. The limit TensorFlow is hitting will be less than that. To see the limit I would suggest looking at this section of your out of memory error:
```
2020-04-02 12:59:38.768619: I tensorflow/core/common_runtime/bfc_allocator.cc:1461] Stats: 
Limit:                         31622483200
InUse:                         31397388288
MaxInUse:                      31399995392
NumAllocs:                            2298
MaxAllocSize:                   3691777280
BytesInactive:                           0
BytesActive:                   31397388288
PeakBytesActive:               31399995392
```
The Limit and InUse values would useful to know in this case.

In email you mentioned you are calling fit multiple times, once per epoch and you hit the out of memory error after the first epoch, when calling fit a second time.

After the first epoch it may be useful to call `print('Reclaimed bytes: ', tf.experimental.get_bytes_reclaimed(0))` to see how many bytes LMS moved from GPU to system memory in the first epoch. (See: https://github.com/IBM/tensorflow-large-model-support/blob/master/examples/AllocatorStats.md)

Also, instead of calling fit multiple times, could you pass `epochs=` to the `fit` call and see if that makes a difference?

**Comment by smatzek** [comment]
I just tried calling fit multiple times using the [ManyModel example](https://github.com/IBM/tensorflow-large-model-support/tree/master/examples) and it did not change the peak amount of memory reclaimed per call to fit.

On the topic of exceeding the system memory, it would definitely not be out of the ordinary with large networks and a 16GB system memory. On the LMS for TensorFlow 2 performance blog, https://developer.ibm.com/linuxonpower/2020/03/06/performance-results-with-large-model-support-for-tensorflow-2/, the 3D U-Net model I used was tested with a 32GB V100 GPU on a system with 1TB of system memory. You can see that at 400^3 image size it was swapping out 180GB per batch, which is ~5x the amount of GPU memory (32GB).

Another thing that you could try to test out LMS on your system with its 16GB of RAM and 8GB of GPU memory is to run the ManyModel.py example. The idea would be to run it with increasing sizes of `--image_size` until it runs out of memory. You can then add the `--lms` parameter and continue increasing the image size until you run out of system memory.  Additionally, if you run with the `--steps 15 --lms_stats_average` parameters you can look at output CSV in `./model_outputs` to see how much memory swapping is occurring at each resolution.

**Comment by fcunilim** [comment]
Here is my console output when running my model without LMS enabled:

https://pastebin.com/Uwcux5dG

For convenience, I am listing below the (what I believe are) the most important lines:

```
2020-04-03 10:19:44.194294: W tensorflow/core/common_runtime/bfc_allocator.cc:801] Allocator (GPU_0_bfc) ran out of memory trying to allocate 512.00MiB (rounded to 536870912).  Current allocation summary follows.
2020-04-03 10:19:44.194382: I tensorflow/core/common_runtime/bfc_allocator.cc:1288] BFCAllocator dump for GPU_0_bfc
2020-04-03 10:19:44.194413: I tensorflow/core/common_runtime/bfc_allocator.cc:1295] Bin (256): 	Total Chunks: 120, Chunks in use: 120. 30.0KiB allocated for chunks. 30.0KiB in use in bin. 13.6KiB client-
[...]
2020-04-03 10:19:44.194812: I tensorflow/core/common_runtime/bfc_allocator.cc:1311] Bin for 512.00MiB was 256.00MiB, Chunk State: 
2020-04-03 10:19:44.194827: I tensorflow/core/common_runtime/bfc_allocator.cc:1324] Next region of size 7722755584
2020-04-03 10:19:44.194850: I tensorflow/core/common_runtime/bfc_allocator.cc:1344] InUse  at 7f6982000000 of size 1280 next 1
[...]
2020-04-03 10:19:44.199800: I tensorflow/core/common_runtime/bfc_allocator.cc:1352] 2 Chunks of size 1073741824 totalling 2.00GiB
2020-04-03 10:19:44.199805: I tensorflow/core/common_runtime/bfc_allocator.cc:1356] Sum Total of in-use chunks: 6.88GiB
2020-04-03 10:19:44.199809: I tensorflow/core/common_runtime/bfc_allocator.cc:1358] total_region_allocated_bytes_: 7722755584 memory_limit_: 7722755687 available bytes: 103 curr_region_allocation_bytes_: 15445511680
2020-04-03 10:19:44.199818: I tensorflow/core/common_runtime/bfc_allocator.cc:1364] Stats: 
Limit:                        7722755687
InUse:                        7382963456
MaxInUse:                     7382963456
NumAllocs:                          1406
MaxAllocSize:                 2451767296
BytesInactive:                         0
BytesActive:                  7382963456
PeakBytesActive:              7382963456
BytesReclaimed:                        0
NumSingleReclaims:                     0
NumFullReclaims:                       0
NumDefrags:                            0
BytesDefragged:                        0

2020-04-03 10:19:44.199835: W tensorflow/core/common_runtime/bfc_allocator.cc:806] *********************************************************************_*****************************_
2020-04-03 10:19:44.199845: W tensorflow/core/common_runtime/bfc_allocator.cc:809] Enabling Large Model Support may avoid this failure.
2020-04-03 10:19:44.199863: W tensorflow/core/framework/op_kernel.cc:1676] OP_REQUIRES failed at conv_ops_3d.cc:412 : Resource exhausted: OOM when allocating tensor with shape[1,64,128,128,128] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
2020-04-03 10:19:44.199920: W tensorflow/core/common_runtime/base_collective_executor.cc:217] BaseCollectiveExecutor::StartAbort Resource exhausted: OOM when allocating tensor with shape[1,64,128,128,128] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
	 [[{{node model/conv3d_12/Conv3D}}]]
Hint: If you want to see a list of allocated tensors when OOM happens, add report_tensor_allocations_upon_oom to RunOptions for current allocation info.
```
Note that I left TF_GPU_HOST_MEM_LIMIT_IN_MB undefined, meaning it is taking its default value of 64GB.

EDIT: regarding the last line in my log, how do I set 'report_tensor_allocations_upon_oom'?

**Comment by fcunilim** [comment]
When I enable LMS in my code

- I get the error specified at the beginning of this issue report (see above)
- my computer becomes unresponsive for about 15-25 minutes, the mouse is frozen and the hard drive activity is very high.

I am willing to investigate this issue and learn how to use LMS, but would appreciate if you have a solution for the above problem, which is very difficult to cope with.

> After the first epoch it may be useful to call `print('Reclaimed bytes: ', tf.experimental.get_bytes_reclaimed(0))` to see how many bytes LMS moved from GPU to system memory in the first epoch. (See: https://github.com/IBM/tensorflow-large-model-support/blob/master/examples/AllocatorStats.md)
> 
> Also, instead of calling fit multiple times, could you pass `epochs=` to the `fit` call and see if that makes a difference?

The email I had originally sent you privately prior to migrating my questions to github did mention I managed to run my model once with LMS enabled, and that it did work once for the first epoch.

However, I am no longer able to reproduce this. I tried maybe 10 times (everytime, my computer takes over 15 minutes to recover...) and unfortunately, I never get past the first epoch.

**Comment by fcunilim** [comment]
I just tried again with LMS enabled. I used `TF_GPU_HOST_MEM_LIMIT_IN_MB=10000` (so as to avoid having to freeze my computer) for this test and it **did go beyond the first epoch**.

My log output is the following for this test:
https://pastebin.com/ZC7mJ8pY

>Does your full error message contain a message like this?
> Allocator (gpu_host_bfc) ran out of memory trying to allocate ... Note the gpu_host_bfc allocator. I believe your message will have this in it. The gpu_host_bfc allocator is the allocator of system/CPU memory, not GPU memory.

Yes, I do have a line mentionning gpu_host_bfc:

```
2020-04-03 10:47:38.813254: W tensorflow/core/common_runtime/bfc_allocator.cc:801] Allocator (gpu_host_bfc) ran out of memory trying to allocate 512.00MiB (rounded to 536870912).  Current allocation summary follows.
2020-04-03 10:47:38.813359: I tensorflow/core/common_runtime/bfc_allocator.cc:1288] BFCAllocator dump for gpu_host_bfc
[...]

```

**Comment by smatzek** [comment]
Limiting TensorFlow's pinned host memory area by setting `TF_GPU_HOST_MEM_LIMIT_IN_MB=10000` was the right thing to do. This is reserving ~6GB of memory for your operating system and other software.

By setting 10000, you are making your GPU have an effective memory space of about 18 GiB, and I'm glad it was able to get a bit further into the training.

> EDIT: regarding the last line in my log, how do I set 'report_tensor_allocations_upon_oom'?
This was easier to do when running TensorFlow 1. I worked out how to do it in TensorFlow 2 but opted to not include it in the ManyModel example for TensorFlow 2 because it requires you to use TensorFlow 1.x sessions under the covers for training. In addition, when testing with ResNet50 I enabled this and the overhead of tracking the tensors at the highest resolution caused the run to fail poorly and not show the list of tensors. I don' have the example with this enabled handy but I could attempt to dig it up again.

Your question about `report_tensor_allocations_upon_oom` reminded me about another study I did into the memory usage of channels first vs last. Your model is using "channels last" which can cause it to consume a lot more memory when training on GPUs. If you could change it to use channels first it would likely help. See [this article](https://developer.ibm.com/linuxonpower/2019/06/11/image-data-channel-effects-on-memory-usage-and-performance-with-tensorflow-large-model-support/) for more information about it. In the run I discuss in that article I saw the conversion tensors taking ~9GiB. Note that this article was written last year and I was using TensorFlow 1.x. It has links and code examples on how to enable showing tensors on OOM and disabling the layout optimizer that may not work exactly like that on TensorFlow 2.

In your output you may notice this message `Enabling Large Model Support may avoid this failure.` This message is erroneously coming out for the CPU/system RAM allocator and that bug is fixed in the TensorFlow 2.2.0  version we are working on.

Lastly, your output contains ` OP_REQUIRES failed at gpu_swapping_kernels.cc:43 : Resource exhausted: OOM` and `node swap_out_ReluGrad_1_1`. These tokens are not from LMS. They are from TensorFlow's grappler optimizer. In previous, graph-mode/TensorFlow 1.x versions of LMS we disabled this optimizer in WML CE builds and recommended it be [disabled when using non-WML CE builds](https://github.com/IBM/tensorflow-large-model-support/tree/tflmsv1#tensorflow-grappler-and-tensorflow-large-model-support). I did not include this recommendation for TensorFlow 2 because we did not see this memory optimizer interfering with LMS in our testing. In your case however it is likely interfering and attempting a move of memory from GPU to CPU RAM and pushing it over the 10GB capped limit.

I will look into the proper way to disable this in TensorFlow 2 today.

So in summary I suggest:
1. Changing your model to do channels first, or disabling the layout optimizer if you can't. (I will also look into the proper way to disable the optimizer).
2. Slightly increasing the 10GB limit you placed on system memory allocation to see if you can get by the current issue.

**Comment by smatzek** [comment]
TensorFlow 2 has surfaced some of the optimizer settings in the tf.config APIs:
https://www.tensorflow.org/api_docs/python/tf/config/optimizer/set_experimental_options

As a quick immediate test, you could try disabling the layout optimizer like this:
`tf.config.optimizer.set_experimental_options({'layout_optimizer': False})`
This will hopefully reduce your total memory usage as the transpose/layout optimizer tensors will not be consuming memory.

TensorFlow did not expose tf.config settings to modify the memory optimizer setting directly, which is making you hit this error `gpu_swapping_kernels.cc:43`. Disabling the layout optimizer (above) may remove the need to modify this. I will continue to look into a way to disable memory optimizer which may mean running in v1.compat mode and setting the ConfigProto directly like this:  https://github.com/IBM/tensorflow-large-model-support#tensorflow-keras-directly-setting-session


Setting `{'disable_meta_optimizer': True}` would disable it but it would disable all the other optimizations.

**Comment by smatzek** [comment]
I have spent the last several hours running tests with layout optimizer on/off, TensorFlow v1 compat vs TensorFlow 2, and memory optimizer on off.

I ran all combinations of these parameters using a TensorFlow 2.2.0 release candidate with LMS where I have access to a new "peak bytes reclaimed" statistic which helps me see the peak amount of bytes swapped out to system memory. This in turn lets me get a feel for how the different settings affect model usage. I ran the tests using a 32GB V100 on a system that was not system memory constrained (it has 1TB).

I ran all combinations with the ManyModel example using ResNet50 at image resolution 8000x8000. My findings were that with this model, default TensorFlow 2 execution mode swaps the least amount of GPU memory to system memory, which is your current limitation.

I then tried a 3D U-Net using [this model](https://github.com/smatzek/3DUnetCNN). The findings were different here. I found that running in v1 compat mode with default settings swapped less than normal TensorFlow 2 mode. The peak amount of memory swapped out from the GPU was about 80% the peak when running with regular TensorFlow 2 native/eager mode.  I then set the built in (non-LMS) memory optimizer which triggered your failure to NOT swap and LMS swapped even less - about 5% less than with the memory optimizer swapping on.

So those are things you could try. The settings I used to test with v1 compat mode are here:
```python
#import tensorflow as tf

# The below code is to use TensorFlow 2 in v1 compat mode with
# all tf2 behavior disabled and set session configuration variables
# to enable LMS and tweak graph re-writer options.
# To use this, comment out the import tensorflow as tf code above.

import tensorflow.compat.v1 as tf
tf.disable_v2_behavior()
from tensorflow.core.protobuf import rewriter_config_pb2

session_config = tf.compat.v1.ConfigProto(allow_soft_placement=True)
# Enable LMS
session_config.gpu_options.experimental.lms_enabled = True

# Uncommenting this configuration option disables the memory optimizer's swapping
# this showed a good improvement in the amount swapped.
#session_config.graph_options.rewrite_options.memory_optimization = rewriter_config_pb2.RewriterConfig.SCHEDULING_HEURISTICS

# Uncommenting this configuration option disables the layout optimizer
# (see https://developer.ibm.com/linuxonpower/2019/06/11/image-data-channel-effects-on-memory-usage-and-performance-with-tensorflow-large-model-support/)
#session_config.graph_options.rewrite_options.layout_optimizer=rewriter_config_pb2.RewriterConfig.OFF

sess = tf.Session(config=session_config)
tf.keras.backend.set_session(sess)
```

**Comment by fcunilim** [comment]
I tried again this morning and this time, it did work. I ran the test with `TF_GPU_HOST_MEM_LIMIT_IN_MB=10000` and a crop size of 128. It ran for 90+ epochs, so I decided to stop to do some more tests. Good thing! I don't know what happened before. I updated the kernel to 5.3.0-45, so had to reinstall the driver. Weird that this simple thing changed things in any way.

My goal is to go beyond sizes of 128. A crop size of 136 doesn't work unfortunately, even with `TF_GPU_HOST_MEM_LIMIT_IN_MB=12000` (already making my computer unresponsive for a minute or so). Note that, in this case, the log displays the message "Enabling LMS may solve this issue", with all the InUse info before etc.

Now I want to study optimization flags in TF. So I set `TF_GPU_HOST_MEM_LIMIT_IN_MB=6000` and go back to crop size = 128. Of course it doesn't work right away. When disabling both the layout and meta optimizer, it does work though.

I have two remarks/questions.

I would like to try to reduce memory usage. You suggest going from NHWC to NCHW, however in my case (which is 3D), this is more like going from NDHWC to NCDHW. Will the tf.keras Conv3D() work with NDHWC ? (kernel size = list of 3 integers)

If I understand correctly if, my model consumes, say, 13GB, it means that TF+LMS **will** basically use 13GB of pinned central memory. This means that I do need to expand my central memory no matter what (I have 16GB of CPU RAM, and 3GB is not enough to work with). Am I correct? If this is true, this is ok, but I need to order a few simms of RAM.

**Comment by fcunilim** [comment]
Bytes reclaimed after the first epochs (crop size 128, LMS on, MEM_LIMIT=10000):

5790495436
8474849996
11159204556
13843559116
16527913676
19212268236
21896622796
24580977356
etc.

**Comment by smatzek** [comment]
The case where it fails on system out of memory (OOM) and displays the "Enabling LMS may solve this issue" message is a bug in that the message should not show in that case since LMS will not help with system memory outage. I have fixed that in the TensorFlow 2.2.0 level of LMS.

I believe the Conv3D works with both NCDHW and NDHWC. Your model is currently NDHWC since channels are last. The model I use ([Fabian Isensee's architecture as implemented by Ellisdg](https://github.com/smatzek/3DUnetCNN)), does NCDHW and works as well. Here is the start of the `model.summary(line_length=128)` from my model at 128^3:

```
Layer (type)                              Output Shape                Param #        Connected to                               
================================================================================================================================
input_1 (InputLayer)                      [(None, 4, 128, 128, 128)]  0                                                         
________________________________________________________________________________________________________________________________
conv3d (Conv3D)                           (None, 16, 128, 128, 128)   1744           input_1[0][0]                              
________________________________________________________________________________________________________________________________
instance_normalization (InstanceNormaliza (None, 16, 128, 128, 128)   32             conv3d[0][0]                               
________________________________________________________________________________________________________________________________
leaky_re_lu (LeakyReLU)                   (None, 16, 128, 128, 128)   0              instance_normalization[0][0]               
________________________________________________________________________________________________________________________________
conv3d_1 (Conv3D)                         (None, 16, 128, 128, 128)   6928           leaky_re_lu[0][0]                          

```

If your model requires 13GB to train successfully through an iteration or epoch, it will take about 8GB of memory on your GPU and the remainder from system memory. This is an over simplification but is mostly true. LMS will swap out the tensors as needed from the GPU to make room for new allocations. This system memory should then be released/freed at the end of the iteration for tensors that have a lifespan of a single iteration.

The reclaimed bytes statistic function you are calling reports the number of bytes reclaimed since the start of the process. It is a cumulative number so in the later epochs it shows LMS is swapping 2.7 GB per epoch.

**Comment by fcunilim** [comment]
I am closing this issue as you have answered my questions. Sorry I took me so long to reply.
